### PR TITLE
Improve and add tests for parser.

### PR DESCRIPTION
**Pull request by nothingnesses**
Ok so the major changes in this PR so far are:
* I removed some functions that seemed unused/unneeded:
  * `equal_at`, whose usages I replaced with `rest`.
  * `flatten`, which is just the `concat` method for slices.
  * `lines`, which was kinda bad anyway, as it unnecessarily allocated memory.
  * `head_default`, no longer needed due to changes to `skip_spaces`.
* `skip` no longer recurses, which should prevent that function from causing any stack overflows. It also mutates its argument now (not sure if that's bad, doesn't seem like it affects anything though).
* Similarly, the function signatures of `skip_comment` and `skip_spaces` now explicitly state that the argument is mutated (it seems like they were previously implicitly mutating their arguments anyway?).
* The `rest` method now doesn't directly index into a slice (which is dangerous and can cause panics), it instead calls `get` and returns an Option.
* `highlight` should allocate less memory now (only the allocation of the intermediate `String` should be occurring now, instead of also allocating a `Vec`).
* Added tests for a few of the functions (`tail`, the skipper functions, and `highlight`).

### CODE REVIEW

**Comment by VictorTaelin** [comment]
Looks great! But have you tested the new highlight for different inputs? One line, zero lines, 10+ lines, errors covering 1 and multiple lines?

**Comment by nothingnesses** [comment]
Not yet. I'm in the process of writing some property-based tests for functions in parser.rs, which I'm writing in conjunction with this (but that will be another PR), which is why this is still marked as WIP, since I might discover other things to improve in that file while writing the tests and getting a better understanding of its functions.

**Comment by VictorTaelin** [comment]
But are you sure it is the old algorithm? Any chance something changed unnoticed? Just to be sure before merging.

**Comment by nothingnesses** [comment]
It's entirely possible I broke something. I'll definitely do the tests you asked and review any changes before I mark this ready for review.

File: src/parser.rs [hunk 1]
```diff
@@ -379,63 +373,72 @@ pub fn highlight(from_index: usize, to_index: usize, code: &str) -> String {
...
+  for (i, c) in code
+    .chars()
+    .enumerate()
+    .filter(|(_, c)| c == &'\n')
+    .take_while(|(i, _)| i < &to_index) {
       if i < from_index {
         from_line += 1;
       }
       if i < to_index {
         to_line += 1;
```
Comment (nothingnesses) [inline src/parser.rs:385]: TODO: Remove this redundant check.

**Comment by VictorTaelin** [comment]
okay, just let me know when you're confident it is equivalent, and I'll merge!

**Comment by nothingnesses** [comment]
I want to take a break from working on the parser (I'm not sure what other easy improvements I can make to it other than adding more tests) and to instead try to dabble with other parts of the repo, so I've marked this ready for review so that the work I've done so far at least won't be wasted.

**Comment by VictorTaelin** [comment]
I've been short of time to review, test and merge PRs, sorry. Please just notify me via DM on Discord when you need me so merge something. Can't merge this one due to conflicts sadly.

**Pull request merged**